numpy/lib/function_base.py

```diff
@@ -1542,15 +1565,88 @@ def gradient(f, *varargs, **kwargs):
+    [array([[ 1. ,  1. , -0.5],
+            [ 1. ,  1. , -0.5]]), array([[ 2. ,  2. ,  2. ],
+            [ 2. ,  1.7,  0.5]])]
+
+    It is possible to specify how boundaries are treated using `edge_order`
+    
     >>> x = np.array([0, 1, 2, 3, 4])
-    >>> y = x**2
-    >>> np.gradient(y, edge_order=2)
+    >>> f = x**2
+    >>> np.gradient(f, edge_order=1)
+    array([ 1.,  2.,  4.,  6.,  7.])
+    >>> np.gradient(f, edge_order=2)
     array([-0.,  2.,  4.,  6.,  8.])
 
-    The axis keyword can be used to specify a subset of axes of which the gradient is calculated
+    The `axis` keyword can be used to specify a subset of axes of which the gradient is calculated
+    
     >>> np.gradient(np.array([[1, 2, 6], [3, 4, 5]], dtype=np.float), axis=0)
     array([[ 2.,  2., -1.],
            [ 2.,  2., -1.]])
+
+    Notes
+    -----
+    Assuming that :math:`f\\in C^{3}` (i.e., :math:`f` has at least 3 continous derivatives) and 
+    let be :math:`h_{*}` a non homogeneous stepsize, the spacing the finite difference 
+    coefficients are computed by minimising the consistency error :math:`\\eta_{i}`:
+    
+    .. math:: 
+        
+        \\eta_{i} = f_{i}^{\\left(1\\right)} - 
+                    \\left[ \\alpha f\\left(x_{i}\\right) + 
+                            \\beta f\\left(x_{i} + h_{d}\\right) +
+                            \\gamma f\\left(x_{i}-h_{s}\\right)
+                    \\right]
+    
+    By substituting :math:`f(x_{i} + h_{d})` and :math:`f(x_{i} - h_{s})` with their Taylor series
+    expansion, this translates into solving the following the linear system:
+
+    .. math:: 
+
+        \\left\\{ 
+            \\begin{array}{r}
+                \\alpha+\\beta+\\gamma=0 \\\\
+                -\\beta h_{d}+\\gamma h_{s}=1 \\\\
+                \\beta h_{d}^{2}+\\gamma h_{s}^{2}=0
+            \\end{array}
+        \\right.
+
+    The resulting approximation of :math:`f_{i}^{(1)}` is the following:
+
+    .. math:: 
+    
+        \\hat f_{i}^{(1)} = 
+            \\frac{ 
+                h_{s}^{2}f\\left(x_{i} + h_{d}\\right) 
+                + \\left(h_{d}^{2} - h_{s}^{2}\\right)f\\left(x_{i}\\right) 
+                - h_{d}^{2}f\\left(x_{i}-h_{s}\\right)}
+                { h_{s}h_{d}\\left(h_{d} + h_{s}\\right)}
+            + \mathcal{O}\\left(\\frac{h_{d}h_{s}^{2} + h_{s}h_{d}^{2}}{h_{d} + h_{s}}\\right)
+
+    It is worth noting that if :math:`h_{s}=h_{d}` (i.e., data are evenly spaced) 
+    we find the standard second order approximation:
+    
+    .. math:: 
+        
+        \\hat f_{i}^{(1)}=
+            \\frac{f\\left(x_{i+1}\\right) - f\\left(x_{i-1}\\right)}{2h} 
+            + \mathcal{O}\\left(h^{2}\\right)
+
+    With a similar procedure the forward/backward approximations used for boundaries can be derived.
+    
+    References
+    ----------
+    .. [1]  Quarteroni A., Sacco R., Saleri F. (2007) Numerical Mathematics (Texts in Applied Mathematics). New York: Springer.
```

Comment:
The lines should be broken to keep them < 80 characters. It is OK to do that for references as long as you keep the indentation.